File: frontend/lib/api/types/convert.ts
/**
 * Conversion, clipping, repo, and url fetch API types.
 */

export interface ConvertOptions {
  recursive?: boolean;
  extensions?: string[];
  strip_headers_footers?: boolean;
  write_images?: boolean;
  pages?: string;
}

export interface ConvertItem {
  file_id: string;
  name: string;
  output_name?: string;
  output_size?: number;
  source_tokens: number;
  target_tokens: number;
  percent: number;
  output_file_id?: string;
  status?: 'queued' | 'converting' | 'done' | 'skipped' | 'error';
  error?: string;
}

export interface ConvertResponse {
  results: ConvertItem[];
  converted_count: number;
  failed_count: number;
  total_source_tokens: number;
  total_target_tokens: number;
  total_percent: number;
}

export interface ConvertRequest {
  session_id: string;
  file_ids: string[];
  options?: ConvertOptions;
}

export interface ClipOptions {
  write?: boolean;
  strip_headers_footers?: boolean;
  write_images?: boolean;
  pages?: string;
}

export interface ClipRequest {
  session_id: string;
  file_ids: string[];
  options?: ClipOptions;
}

export interface ClipResponse {
  text: string;
  chars: number;
  lines: number;
  tokens: number;
  file_count: number;
}

export interface FetchRequest {
  url: string;
  session_id?: string;
  user_agent?: string;
}

export interface FetchResponse {
  session_id?: string;
  output_file_id: string;
  output_name: string;
  source_tokens: number;
  target_tokens: number;
  percent: number;
  url: string;
  title?: string;
  text?: string;
}

export interface RepoRequest {
  session_id: string;
  file_ids: string[];
  exclude?: string[];
}

export interface RepoResponse {
  output_file_id: string;
  output_name: string;
  source_tokens: number;
  target_tokens: number;
  percent: number;
  file_count: number;
}

File: backend/schema_models/convert.py
"""Pydantic models for single-file conversion and clipboard operations."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ConvertOptions(BaseModel):
    """Conversion options shared by convert / merge / clip / watch."""

    strip_headers_footers: bool = False
    write_images: bool = False
    image_path: str | None = None
    pages: list[int] | None = None
    extensions: list[str] | None = None


class ConvertItem(BaseModel):
    """Per-file result of a convert job."""

    file_id: str
    name: str
    status: str
    output_file_id: str | None = None
    output_name: str | None = None
    output_size: int | None = None
    source_tokens: int = 0
    target_tokens: int = 0
    percent: float = 0.0
    error: str | None = None


class ConvertResponse(BaseModel):
    """Aggregated result of a convert job."""

    results: list[ConvertItem]
    converted_count: int
    failed_count: int
    total_source_tokens: int
    total_target_tokens: int
    total_percent: float


class ConvertRequest(BaseModel):
    session_id: str
    file_ids: list[str] = Field(default_factory=list)
    options: ConvertOptions = Field(default_factory=ConvertOptions)
    path: str | None = None


class ClipRequest(BaseModel):
    session_id: str
    file_ids: list[str] = Field(default_factory=list)
    options: ConvertOptions = Field(default_factory=ConvertOptions)


class ClipResponse(BaseModel):
    text: str
    chars: int
    lines: int
    tokens: int
    file_count: int
